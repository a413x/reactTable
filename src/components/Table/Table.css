.table table{
  table-layout: fixed;
}
.table td{
  overflow: hidden;
}
@media (max-width: 576px) {
  .table table{
    table-layout: auto;
  }
  .table td{
    overflow: visible;
  }
}
