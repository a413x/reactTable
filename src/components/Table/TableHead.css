.table th{
  position: relative;
  cursor: pointer;
  user-select: none;
  overflow: hidden;
  width: 16%;
}
.table th:first-child{
  width: 8%;
}
.table th:nth-child(0n+4){
  width: 36%;
}
.table th:last-child{
  width: 24%;
}
.asc:after,.desc:after{
  content: '';
  position: absolute;
  top: 20px;
  right: 0;
  border-width: 5px 5px 0;
  border-style: solid;
  border-color: #33a transparent;
}
.desc:after{
  border-width: 0 5px 5px;
}
@media (min-width: 576px) {
  .table th{
    overflow: visible;
  }
}
