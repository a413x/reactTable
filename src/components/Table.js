import {TableHead, tableColumns} from './TableHead.js'

export const Table = ({data, sortCallback, showDetailsCallback}) => {
  const onRowClick = (dataObj) => showDetailsCallback(dataObj)
  return (
    <div className = 'table-container container'>
      <table className = 'table table-hover'>
        <TableHead sortCallback = {sortCallback} />
        <tbody>
          {data.map((dataObj,ind) =>
            <tr key = {'tr-' + ind} onClick = {(e) => onRowClick(dataObj)}>
              {tableColumns.map((colObj, ind) =>
                <td key = {ind}>
                  {dataObj[colObj.name]}
                </td>
              )}
            </tr>
          )}
        </tbody>
      </table>
    </div>
  )
}
