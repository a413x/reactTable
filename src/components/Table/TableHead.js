import {useState} from 'react'
import './TableHead.css'

export const tableColumns = [
  {name: 'id', title: '#'},
  {name: 'firstName', title: 'First name'},
  {name: 'lastName', title: 'Last name'},
  {name: 'email', title: 'E-mail'},
  {name: 'phone', title: 'Phone'},
]

export const TableHead = ({sortCallback}) => {
  const [activeCol, setActiveCol] = useState({name: '', order: 'asc'})

  const onClick = (colObj) => {
    let order = activeCol.order
    if(activeCol.name === colObj.name){
      order = order === 'asc' ? 'desc' : 'asc'
    }else{
      order = 'asc'
    }
    setActiveCol({name: colObj.name, order: order})
    sortCallback(colObj, order)
  }

  return (
    <thead>
      <tr>
        {tableColumns.map((colObj, ind) =>
          <th
            key = {'th-' + ind}
            onClick = {() => onClick(colObj)}
            className = {
              activeCol.name === colObj.name ? activeCol.order : ''
            }
          >
            {colObj.title}
          </th>
        )}
      </tr>
    </thead>
  )
}
