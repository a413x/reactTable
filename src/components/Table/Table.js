import {TableHead, tableColumns} from './TableHead.js'
import './Table.css'

export const Table = ({data, sortedCol, sortCallback, showDetailsCallback}) => {
  const onRowClick = (dataObj) => showDetailsCallback(dataObj)
  return (
    <div className = 'table'>
      <table className = 'table table-hover'>
        <TableHead sortCallback = {sortCallback} sortedCol = {sortedCol} />
        <tbody>
          {data.map((dataObj,ind) =>
            <tr key = {'tr-' + ind} onClick = {(e) => onRowClick(dataObj)}>
              {tableColumns.map((colObj, ind) =>
                <td key = {ind}>
                  {dataObj[colObj.name]}
                </td>
              )}
            </tr>
          )}
        </tbody>
      </table>
    </div>
  )
}
