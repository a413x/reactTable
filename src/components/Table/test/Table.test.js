import {render, fireEvent} from '@testing-library/react'
import {Table} from '../Table.js'

describe('Table component tests', () => {
  const mock_sortCallback = jest.fn()
  const mock_showDetailsCallback = jest.fn()
  const props = {
    data: [...mock_data],
    sortCallback: mock_sortCallback,
    showDetailsCallback: mock_showDetailsCallback
  }

  test('renders properly', () => {
    const {container} = render(<Table {...props}/>)
    expect(container).toMatchSnapshot()
  })
  test('showDetailsCallback function calls on table row click', () => {
    render(<Table {...props}/>)
    const index = 2
    const row = document.querySelectorAll('tbody tr')[index]
    fireEvent.click(row)
    expect(mock_showDetailsCallback).toHaveBeenCalledWith(mock_data[index])
  })
})
