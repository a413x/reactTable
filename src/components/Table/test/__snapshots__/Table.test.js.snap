// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component tests renders properly 1`] = `
<div>
  <div
    class="table"
  >
    <table
      class="table table-hover"
    >
      <thead>
        <tr>
          <th
            class=""
          >
            #
          </th>
          <th
            class=""
          >
            First name
          </th>
          <th
            class=""
          >
            Last name
          </th>
          <th
            class=""
          >
            E-mail
          </th>
          <th
            class=""
          >
            Phone
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            Frederick
          </td>
          <td>
            Catania
          </td>
          <td>
            GHellems@magna.net
          </td>
          <td>
            (458)969-6847
          </td>
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            Curtis
          </td>
          <td>
            Bias
          </td>
          <td>
            CRobichaud@nec.gov
          </td>
          <td>
            (481)075-5034
          </td>
        </tr>
        <tr>
          <td>
            3
          </td>
          <td>
            Dora
          </td>
          <td>
            Crenshaw
          </td>
          <td>
            MSanchez@at.org
          </td>
          <td>
            (959)257-7541
          </td>
        </tr>
        <tr>
          <td>
            4
          </td>
          <td>
            Grace
          </td>
          <td>
            French
          </td>
          <td>
            JLund@vestibulum.com
          </td>
          <td>
            (976)555-7134
          </td>
        </tr>
        <tr>
          <td>
            5
          </td>
          <td>
            Randy
          </td>
          <td>
            Bunch
          </td>
          <td>
            KGunter@lacus.io
          </td>
          <td>
            (941)539-5519
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
